spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/procedures/GenerateChangesProcedure.java

```diff
@@ -0,0 +1,336 @@
+    }
+    return viewName;
+  }
+
+  private Dataset<Row> withUpdate(Dataset<Row> df, String[] identifiers) {
+    Column[] partitionSpec = getPartitionSpec(df, identifiers);
+    Column[] sortSpec = sortSpec(df, partitionSpec);
+
+    int changeTypeIdx = df.schema().fieldIndex(MetadataColumns.CHANGE_TYPE.name());
+    List<Integer> partitionIdx =
+        Arrays.stream(partitionSpec)
+            .map(column -> df.schema().fieldIndex(column.toString()))
+            .collect(Collectors.toList());
+
+    return df.repartition(partitionSpec)
+        .sortWithinPartitions(sortSpec)
+        .mapPartitions(
+            processRowsWithinTask(changeTypeIdx, partitionIdx), RowEncoder.apply(df.schema()));
+  }
+
+  @NotNull
+  private static Column[] getPartitionSpec(Dataset<Row> df, String[] identifiers) {
+    Column[] partitionSpec = new Column[identifiers.length + 1];
+    for (int i = 0; i < identifiers.length; i++) {
+      try {
+        partitionSpec[i] = df.col(identifiers[i]);
+      } catch (Exception e) {
+        throw new IllegalArgumentException(
+            String.format("Identifier column '%s' does not exist in the table", identifiers[i]), e);
+      }
+    }
+    partitionSpec[partitionSpec.length - 1] = df.col(MetadataColumns.CHANGE_ORDINAL.name());
+    return partitionSpec;
+  }
+
+  @NotNull
+  private static Column[] sortSpec(Dataset<Row> df, Column[] partitionSpec) {
+    Column[] sortSpec = new Column[partitionSpec.length + 1];
+    System.arraycopy(partitionSpec, 0, sortSpec, 0, partitionSpec.length);
+    sortSpec[sortSpec.length - 1] = df.col(MetadataColumns.CHANGE_TYPE.name());
+    return sortSpec;
+  }
+
+  private static MapPartitionsFunction<Row, Row> processRowsWithinTask(
+      int changeTypeIndex, List<Integer> partitionIdx) {
+    return rowIterator -> {
+      Iterator<Row> iterator =
+          new Iterator<Row>() {
+            private Row nextRow = null;
+
+            @Override
+            public boolean hasNext() {
+              if (nextRow != null) {
+                return true;
+              }
+              return rowIterator.hasNext();
+            }
+
+            @Override
+            public Row next() {
+              // if the next row is cached and changed, return it directly
+              if (nextRow != null
+                  && !nextRow.getString(changeTypeIndex).equals(DELETE.name())
+                  && !nextRow.getString(changeTypeIndex).equals(INSERT.name())) {
+                Row row = nextRow;
+                nextRow = null;
+                return row;
+              }
+
+              Row currentRow = currentRow();
+
+              if (rowIterator.hasNext()) {
+                GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+
+                if (withinPartition(currentRow, nextRow)
+                    && currentRow.getString(changeTypeIndex).equals(DELETE.name())
+                    && nextRow.getString(changeTypeIndex).equals(INSERT.name())) {
```

Comment:
It is rare that two rows are from different partitions, but it can happen, especially AQE can merge different partitions together. It is hard to add a unit test for this case though.

Reply:
I think creating a separate class for this functionality will allow us to unit test it.

Reply:
Done and added unit tests.